### PR TITLE
Prevent pumping waits in more places in AbstractProject

### PR DESCRIPTION
**Pull request by Pilchie**
Specifically, prevent it in OnDocumentOpened, OnDocumentClosed, and SetOptions.
Windows error reports show crashes due to reentrancy on re-opening web projects
where a document is being re-opened, and we pump while acquiring a lock in the
workspace inside and OnDocumentOpened() call.  During the pumping, the ASP.NET
async parse request finishes, and they call back into the project to
SetOptions(), which calls into the Workspace and we crash with a
LockRecursionException.  I added the using in OnDocumentClosed() for symmetry,
though I don't have concrete proof of a codepath that re-enters through it yet.

Fixes internal issue 179940.

### CODE REVIEW

**Comment by Pilchie** [comment]
test vsi please

**Comment by Pilchie** [comment]
@dotnet/roslyn-ide Would love if someone could take a look at this - especially you @jasonmalinowski.  Also @mattwar.

**Comment by Pilchie** [comment]
I manually tested the repro scenario of re-opening a VB Web Application with a file open in the .suo, and it worked fine, but this was a timing issue and I could never repro the crash without the fix either :-/

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs [hunk 1]
```diff
@@ -636,28 +640,36 @@ protected void RemoveProjectReference(ProjectReference projectReference)
...
-            IVisualStudioHostDocument document = (IVisualStudioHostDocument)sender;
-            AbstractProject project = (AbstractProject)document.Project;
-
-            if (project._pushingChangesToWorkspaceHosts)
-            {
-                project.ProjectTracker.NotifyWorkspaceHosts(host => host.OnDocumentOpened(document.Id, document.GetOpenTextBuffer(), isCurrentContext));
-            }
-            else
+            // We do not want to allow message pumping/reentrancy when processing project system changes.
+            using (Dispatcher.CurrentDispatcher.DisableProcessing())
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs]: My impression was StartPushingToWorkspaceAndNotifyOfOpenDocuments and NotifyWorkspaceHosts already did this. Should we just be pushing this to there?

Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs]: Indeed they do...did that not already work, or did we fix that post Update 1?

Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs]: Indeed we did fix that after update 2 as part of a0bfff to fix this very same bug.  Wonder why I didn't resolve it then.

**Comment by Pilchie** [comment]
Already fixed in `VisualStudioProjectTracker.NotifyWorkspaceHosts`.

**Pull request closed without merging**